www/ap2-authn-otp/work/mod_authn_otp-1.1.3/otptool.1
.\"  -*- nroff -*-
.\"
.\" otptool - HOTP/OATH one-time password utility
.\"
.\" Copyright 2009 Archie L. Cobbs <archie@dellroad.org>
.\"
.\" Licensed under the Apache License, Version 2.0 (the "License");
.\" you may not use this file except in compliance with the License.
.\" You may obtain a copy of the License at
.\"
.\"     http://www.apache.org/licenses/LICENSE-2.0
.\"
.\" Unless required by applicable law or agreed to in writing, software
.\" distributed under the License is distributed on an "AS IS" BASIS,
.\" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\" See the License for the specific language governing permissions and
.\" limitations under the License.
.\"
.\" $Id: otptool.1,v 1.1.1.1 2011/04/21 09:49:46 pettai Exp $
.\"/
.Dd June 21, 2008
.Dt OTPTOOL 1
.Os
.Sh NAME
.Nm otptool
.Nd HOTP/OATH one-time password utility
.Sh SYNOPSIS
.Nm otptool
.Bk -words
.Op Fl fht
.Op Fl c Ar counter
.Op Fl d Ar #digits
.Op Fl i Ar secs
.Op Fl m Ar PIN
.Op Fl w Ar num
.Ar key
.Op Ar password
.Ek
.Sh DESCRIPTION
.Nm
is a utility for generating, verifying, and synchronizing one-time passwords
created using the HOTP/OATH algorithm defined by RFC 4226.
.Pp
The
.Ar key
is the token's binary secret key and is specified as a hexadecimal string.
.Pp
If no
.Ar password
is given,
.Nm
generates the one-time password corresponding to the given key and target counter value
and prints to standard output the counter followed by the decimal and hexadecimal one-time passwords.
If
.Ar password
is given, then
.Nm
verifies that
.Ar password
is the correct one-time password for the given
.Ar key
and counter value.
If so, it outputs the counter value.
.Ar password
may be either the decimal or hexadecimal one-time password.
.Pp
The target counter value is determined as follows: if the
.Fl t
flag is given, use the current time in seconds since the UNIX epoch divided by the configured time interval (default 30 seconds);
otherwise, if the
.Fl c
flag is given, use the given
.Ar counter ;
otherwise, use the value zero.
.Pp
In both cases, a range of target counter values may be specified using the
.Fl w
flag.
When both
.Fl w
and
.Ar password
are given,
.Nm
will search the entire range for a matching counter value,
starting with the target counter value and working away from it.
This mode can be used to resynchronize an unsychronized counter.
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl c
Specify the starting target counter value for the one-time password generation or search.
This flag is incompatible with the
.Fl t
flag;
if neither flag is given, the default value is zero.
.It Fl d
Specify the required number of digits in the one-time password.
Giving a
.Ar password
argument and specifying a different length here will result in no match being found (no search is performed).
Otherwise, the default value is the length of
.Ar password ,
if given, or else six if not.
.It Fl f
Read the key from the file named
.Ar key
instead of parsing
.Ar key
as a hexadecimal string.
.It Fl h
Print the usage message and exit successfully.
.It Fl i
Specify the length of a single time interval in seconds.
The default value is 30 seconds.
This flag is ignored unless the
.Fl t
flag is also given.
.It Fl m
Use the Mobile-OTP algorithm with the given PIN instead of the HOTP/OATH algorithm.
This flag imples
.Fl i Ar 10
and
.Fl d Ar 6 .
Normally you also want to specify
.Fl t .
.It Fl t
Use the current time as the basis for the target counter value.
This flag is incompatible with the
.Fl c
flag.
.It Fl w
Specify the width of a window of counter values within which to iterate when
generating or searching for one-time passwords.
When
.Fl t
is used, the window extends the given distance both before and after the target counter value;
otherwise, the window extends forward of the target counter value.
When both
.Ar password
and
.Fl t
are given, the search starts with the initial target counter and works away from it
in both directions.
.El
.Sh RETURN VALUE
.Nm
exits with one of the following return values:
.Bl -tag -width xxx
.It 0
The one-time password(s) was/were successfully generated, or
.Ar password
correctly matched the password generated using (one of) the target counter value(s).
.It 1
.Nm
was invoked with invalid command line flags or parameters.
.It 2
The given
.Ar password
did not match any counter value(s) in the search window.
.It 3
A system error occurred.
.El
.Sh SEE ALSO
.Rs
.%T "HOTP: An HMAC-Based One-Time Password Algorithm"
.%O "http://www.ietf.org/rfc/rfc4226.txt"
.Re
.Rs
.%T "mod_authn_otp: Apache module for one-time password authentication"
.%O "http://mod-authn-otp.googlecode.com/"
.Re
.Rs
.%T "Mobile-OTP: Mobile One Time Passwords"
.%O "http://motp.sourceforge.net/"
.Re
.Sh AUTHOR
.An Archie L. Cobbs Aq archie@dellroad.org


doc/NEWS
# $NetBSD: NEWS,v 1.5 2008/04/13 13:24:16 tnn Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

buildlink3.mk layout changed [joerg 2009-03-20]
	The buildlink3.mk files have been reorganized to reduce
	redundant parsing for deeply nested buildlink3 trees.
	The createbuildlink package has been updated to create
	the new format.

PLIST_VARS introduced [jlam 20080413]
	PLIST_VARS is a simpler way to do conditional PLIST entries.
	The old method of doing e.g. PLIST_SUBST+= FOO="@comment " can
	now be done by unconditionally setting PLIST_VARS+=foo and
	conditionally setting PLIST.foo=yes. This will result in substitution
	of ${PLIST.foo} in the PLIST with either "" or "@comment ".

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


www/ap2-authn-otp/work/mod_authn_otp-1.1.3/otptool.c

/*
 * otptool - HOTP/OATH one-time password utility
 *
 * Copyright 2009 Archie L. Cobbs <archie@dellroad.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * $Id: otptool.c,v 1.1.1.1 2011/04/21 09:49:46 pettai Exp $
 */

#include "otptool.h"

/* Definitions */
#define OTP_BUF_SIZE       16

/* Internal functions */
static void         usage(void);

int
main(int argc, char **argv)
{
    const char *otp = NULL;
    const char *key = NULL;
    const char *motp_pin = NULL;
    unsigned char keybuf[128];
    char otpbuf10[OTP_BUF_SIZE];
    char otpbuf16[OTP_BUF_SIZE];
    size_t keylen;
    int time_interval = DEFAULT_TIME_INTERVAL;
    int ndigits = -1;
    int counter_start;
    int counter_stop;
    int read_from_file = 0;
    int counter = -1;
    int use_time = 0;
    int window = 0;
    int ch;
    int i;

    /* Parse command line */
    while ((ch = getopt(argc, argv, "c:d:fhi:m:tvw:")) != -1) {
        switch (ch) {
        case 'c':
            if (use_time)
                errx(EXIT_USAGE_ERROR, "only one of `-c' or `-t' should be specified");
            counter = atoi(optarg);
            if (counter < 0)
                errx(EXIT_USAGE_ERROR, "invalid counter value `%s'", optarg);
            break;
        case 'd':
            ndigits = atoi(optarg);
            if (ndigits < 1)
                errx(EXIT_USAGE_ERROR, "invalid digit count `%s'", optarg);
            break;
        case 'f':
            read_from_file = 1;
            break;
        case 'h':
            usage();
            return 0;
        case 'i':
            time_interval = atoi(optarg);
            break;
        case 'm':
            ndigits = 6;
            time_interval = 10;
            motp_pin = optarg;
            *otpbuf10 = '\0';
            break;
        case 't':
            if (counter != -1)
                errx(EXIT_USAGE_ERROR, "only one of `-c' or `-t' should be specified");
            use_time = 1;
            break;
        case 'w':
            window = atoi(optarg);
            if (window < 0)
                errx(EXIT_USAGE_ERROR, "invalid counter window `%s'", optarg);
            break;
        default:
            warnx("unrecognized flag `-%c'", ch);
            usage();
            return EXIT_USAGE_ERROR;
        }
    }

    /* Parse command line arguments */
    switch (argc - optind) {
    case 2:
        otp = argv[optind + 1];
        if (ndigits == -1)
            ndigits = strlen(otp);
        if (strlen(otp) != ndigits)
            errx(EXIT_NOT_MATCHED, "the given OTP `%s' has the wrong length %d != %d", otp, (int)strlen(otp), ndigits);
        // FALLTHROUGH
    case 1:
        key = argv[optind];
        break;
    default:
        warnx("wrong number of command line arguments");
        usage();
        return EXIT_USAGE_ERROR;
    }

    /* Set default #digits */
    if (ndigits == -1)
        ndigits = DEFAULT_NUM_DIGITS;

    /* Read or parse key */
    if (read_from_file) {
        FILE *fp;

        if ((fp = fopen(key, "rb")) == NULL)
            err(EXIT_SYSTEM_ERROR, "error reading `%s'", key);
        keylen = fread(keybuf, 1, sizeof(keybuf), fp);
        if (ferror(fp))
            err(EXIT_SYSTEM_ERROR, "error reading `%s'", key);
        if (!feof(fp))
            errx(EXIT_SYSTEM_ERROR, "error reading `%s': %s", key, "key is too long");
        fclose(fp);
    } else {
        for (keylen = 0; keylen < sizeof(keybuf) && key[keylen * 2] != '\0'; keylen++) {
            const char *s = &key[keylen * 2];
            int nibs[2];

            for (i = 0; i < 2; i++) {
                if (isdigit(s[i]))
                    nibs[i] = s[i] - '0';
                else if (isxdigit(s[i]))
                    nibs[i] = tolower(s[i]) - 'a' + 10;
                else
                    errx(EXIT_USAGE_ERROR, "invalid key `%s'", key);
            }
            keybuf[keylen] = (nibs[0] << 4) | nibs[1];
        }
    }

    /* Determine target counter */
    if (use_time)
        counter = time(NULL) / time_interval;
    else if (counter < 0)
        counter = 0;

    /* Search or generate */
    if (otp == NULL) {
        if (use_time) {
            counter_start = counter - window;
            counter_stop = counter + window;
        } else {
            counter_start = counter;
            counter_stop = counter + window;
        }
        for (counter = counter_start; counter <= counter_stop; counter++) {
            if (motp_pin != NULL)
                motp(keybuf, keylen, motp_pin, counter, ndigits, otpbuf16, OTP_BUF_SIZE);
            else
                hotp(keybuf, keylen, counter, ndigits, otpbuf10, otpbuf16, OTP_BUF_SIZE);
            printf("%d: %s%s%s\n", counter, otpbuf10, *otpbuf10 != '\0' && *otpbuf16 != '\0' ? " " : "", otpbuf16);
        }
        return 0;
    } else {
        for (i = 0; i <= window; i++) {
            int try;

            try = counter + i;
            if (motp_pin != NULL)
                motp(keybuf, keylen, motp_pin, try, ndigits, otpbuf16, OTP_BUF_SIZE);
            else
                hotp(keybuf, keylen, try, ndigits, otpbuf10, otpbuf16, OTP_BUF_SIZE);
            if (strcasecmp(otp, otpbuf10) == 0 || strcasecmp(otp, otpbuf16) == 0)
                goto match;
            if (use_time && i != 0) {
                try = counter - i;
                if (motp_pin != NULL)
                    motp(keybuf, keylen, motp_pin, try, ndigits, otpbuf16, OTP_BUF_SIZE);
                else
                    hotp(keybuf, keylen, try, ndigits, otpbuf10, otpbuf16, OTP_BUF_SIZE);
                if (strcasecmp(otp, otpbuf10) == 0 || strcasecmp(otp, otpbuf16) == 0)
                    goto match;
            }
            continue;
match:
            printf("%d\n", try);
            return 0;
        }
    }

    /* Not found */
    fprintf(stderr, "one-time password \"%s\" was not found within the counter range %d ... %d\n", otp,
      use_time ? counter - window : counter, counter + window);
    return EXIT_NOT_MATCHED;
}

static void
usage()
{
    fprintf(stderr, "Usage: %s [-fht] [-c counter] [-d digits] [-i interval] [-m PIN] [-w window] key [otp]\n", PROG_NAME);
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -c\tSpecify the initial counter value (conflicts with `-t')\n");
    fprintf(stderr, "  -f\t`key' refers to the file containing the key\n");
    fprintf(stderr, "  -h\tDisplay this usage message\n");
    fprintf(stderr, "  -i\tSpecify time interval in seconds (default %d)\n", DEFAULT_TIME_INTERVAL);
    fprintf(stderr, "  -m\tUse mOTP algorithm with given PIN; also implies `-d 6' and `-i 10'\n");
    fprintf(stderr, "  -t\tDerive initial counter value from the current time (conflicts with `-c')\n");
    fprintf(stderr, "  -n\tSpecify number of digits in the generated OTP(s) (default %d)\n", DEFAULT_NUM_DIGITS);
    fprintf(stderr, "  -w\tSpecify size of window for additional counter values (default %d)\n", DEFAULT_WINDOW);
}

